provider/src/main/kotlin/org/gradle/kotlin/dsl/resolver/SourcePathProvider.kt

```diff
@@ -40,13 +43,18 @@ object SourcePathProvider {
     }
 
     /**
-     * Returns all conventional source directories under buildSrc if any.
-     *
-     * This won't work for buildSrc projects with a custom source directory layout
-     * but should account for the majority of cases and it's cheap.
+     * Returns source directories from buildSrc if any.
      */
     private
     fun buildSrcRootsOf(projectRoot: File): Collection<File> =
+        projectRoot.resolve("buildSrc/build/$buildSrcSourceRootsFilePath")
```

Comment:
:x: The task is writing to `layout.buildDirectory.file(buildSrcSourceRootsFilePath)` and the provider is reading from `projectRoot.resolve("buildSrc/build/$buildSrcSourceRootsFilePath")` which wouldn't match for custom layouts. Hardcoding `build` on both sides seems slightly better:
1. The IDE experience would not deteriorate given a custom `layout.buildDirectory` in `buildSrc`
2. We are more likely to get accurate bug reports when/if this persistent `buildSrc/build` directory becomes a problem than for IDE features seemingly not working

As an alternative, have you considered embedding the `source-roots.txt` file as a resource?

Reply:
I did initially embed the `source-roots.txt` file as a resource in `buildSrc.jar` but given we don't generate that file on the regular usage, only when in classpath mode, it would have caused too much classpath invalidation. So, I went back to directly pointing at the file.

Otherwise, good point about the `build` path segment, I did what you suggested.

Reply:
What if it is always generated with a cacheable task?

Reply:
This would work. But it would mean adding the project configuration and task creation done by `BuildSrcSourceRootsConfigurationAction` plus `projectDependenciesSourceRootsFrom()` overhead to all builds out there. Execution of `projectDependenciesSourceRootsFrom()` would be required always even when fetching from cache as it's actually the inputs of the task.

Reply:
Makes sense.